File: golang/mathops/math_operations.go
package mathops

import (
	"errors"
	"math"
)

// Add returns the sum of two integers.
func Add(a, b int) int {
	return a + b
}

// Subtract returns the difference of two integers.
func Subtract(a, b int) int {
	return a - b
}

// Multiply returns the product of two integers.
func Multiply(a, b int) int {
	return a * b
}

// Divide returns the quotient of two integers as a float64.
// Returns an error if b is zero.
func Divide(a, b int) (float64, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return float64(a) / float64(b), nil
}

// Power returns base raised to the given exponent.
func Power(base float64, exponent int) float64 {
	return math.Pow(base, float64(exponent))
}

// Factorial returns the factorial of n.
// Panics if n is negative.
func Factorial(n int) int64 {
	if n < 0 {
		panic("factorial is not defined for negative numbers")
	}
	if n == 0 || n == 1 {
		return 1
	}
	var result int64 = 1
	for i := 2; i <= n; i++ {
		result *= int64(i)
	}
	return result
}

// Derivative computes the derivative of a polynomial represented by its
// coefficients. The index of each coefficient corresponds to the power of x.
// For example, [3, 2, 1] represents 3 + 2x + x² and its derivative is [2, 2].
func Derivative(coefficients []float64) []float64 {
	if len(coefficients) <= 1 {
		return []float64{0}
	}
	result := make([]float64, len(coefficients)-1)
	for i := 1; i < len(coefficients); i++ {
		result[i-1] = float64(i) * coefficients[i]
	}
	return result
}

// Pi returns a string representation of pi to 40 decimal places.
func Pi() string {
	return "3.1415926535897932384626433832795028841971"
}

// GCD returns the greatest common divisor of two integers using
// the Euclidean algorithm.
func GCD(a, b int) int {
	if a < 0 {
		a = -a
	}
	if b < 0 {
		b = -b
	}
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

File: csharp/CornTest/MathOperations.cs
namespace CornTest;

/// <summary>
/// Provides deterministic mathematical operations for testing purposes.
/// All methods in this class produce predictable, repeatable results.
/// </summary>
public class MathOperations
{
    /// <summary>Computes the sum of two integers.</summary>
    public int Add(int a, int b) => a + b;

    /// <summary>Computes the difference of two integers.</summary>
    public int Subtract(int a, int b) => a - b;

    /// <summary>Computes the product of two integers.</summary>
    public int Multiply(int a, int b) => a * b;

    /// <summary>
    /// Computes the quotient of two integers as a double.
    /// </summary>
    /// <exception cref="DivideByZeroException">Thrown when the divisor is zero.</exception>
    public double Divide(int a, int b)
    {
        if (b == 0)
            throw new DivideByZeroException("Cannot divide by zero");
        return (double)a / b;
    }

    /// <summary>Raises a base value to the given integer exponent.</summary>
    public double Power(double baseVal, int exponent) => Math.Pow(baseVal, exponent);

    /// <summary>
    /// Computes n! iteratively.
    /// </summary>
    /// <exception cref="ArgumentException">Thrown when n is negative.</exception>
    public long Factorial(int n)
    {
        if (n < 0)
            throw new ArgumentException("Factorial is undefined for negative integers");
        if (n <= 1)
            return 1L;
        long accumulator = 1L;
        for (int idx = 2; idx <= n; idx++)
            accumulator *= idx;
        return accumulator;
    }

    /// <summary>
    /// Computes the derivative of a polynomial whose coefficients are given
    /// in ascending order of degree.
    /// For example, {3, 2, 5} represents 3 + 2x + 5x^2, and the derivative
    /// is {2, 10} representing 2 + 10x.
    /// </summary>
    /// <exception cref="ArgumentNullException">Thrown when coefficients is null.</exception>
    public double[] Derivative(double[] coefficients)
    {
        ArgumentNullException.ThrowIfNull(coefficients);
        if (coefficients.Length <= 1)
            return new double[] { 0.0 };
        var derived = new double[coefficients.Length - 1];
        for (int i = 1; i < coefficients.Length; i++)
            derived[i - 1] = coefficients[i] * i;
        return derived;
    }

    /// <summary>Returns pi to 40 decimal places as a string.</summary>
    public string Pi() => "3.1415926535897932384626433832795028841971";

    /// <summary>
    /// Computes the greatest common divisor via the Euclidean algorithm.
    /// Negative inputs are handled by taking absolute values first.
    /// </summary>
    public int Gcd(int a, int b)
    {
        a = Math.Abs(a);
        b = Math.Abs(b);
        while (b != 0)
        {
            int remainder = a % b;
            a = b;
            b = remainder;
        }
        return a;
    }
}
